jsdom-based-scraping copy/index.md

```diff
@@ -0,0 +1,207 @@
+    preNavigationHooks: [
+        (crawlingContext) => {
+            const { request, session } = crawlingContext;
+            request.headers = {
+                ...request.headers,
+                ...session.userData?.headers,
+
+            };
+        },
+    ],
+    proxyConfiguration,
+```
+Finally in the request handler we make the call using the headers and make sure how many calls are needed to fetch all the data handling pagination.
+
+```js
+async requestHandler(context) {
+        const { log, request, json } = context;
+        const { userData } = request;
+        const { itemsCounter = 0, resultsLimit = 0 } = userData;
+        if (!json.data) {
+            throw new Error('BLOCKED');
+        }
+        const { data } = json;
+        const items = data.list;
+        const counter = itemsCounter + items.length;
+        const dataItems = items.slice(0, resultsLimit && counter > resultsLimit ? resultsLimit - itemsCounter : undefined);
+        await Actor.pushData(dataItems);
+        const { pagination: { page, total } } = data;
+        log.info(`Scraped ${dataItems.length} results out of ${total} from search page ${page}`);
+        const isResultsLimitNotReached = counter < Math.min(total, resultsLimit);
+        if (isResultsLimitNotReached && data.pagination.has_more) {
+            const nextUrl = new URL(request.url);
+            nextUrl.searchParams.set('page', page + 1);
+            await crawler.addRequests([{
+                url: nextUrl.toString(),
+                headers: request.headers,
+                userData: {
+                    ...request.userData,
+                    itemsCounter: itemsCounter + dataItems.length,
+                },
+            }]);
+        }
+    },
+});
+```
+
+In last, one thing important to note is that we want to make any amount of API class. In this example we started with just one request and a single session, but can be any amount. When the first api call is completed it creates second api call, not more since we are expecting all data from two api calls. But if in the future we will need to call other api calls we can do so. 
+
+To make things more clear, here is how code flow looks:
+
+
+![code flow](./images/code-flow.webp)
+
+## Conclusion
+
+This approach helps us to get a third way to extract the authentication data without actually using browser and pass the data to CheerioCrawler. This significantly improves the performance, reduces the RAM requirement by 50%, and while browser based scraping performance is 10 times slower than pure Cheerio, JSDOM do it just 3-4 times slower, which makes it 2-3 times faster than browser based scraping.
+
+We used our open source library Crawlee in this example, feel free to checkout the project and give us a star. The codebase of the project is already [uploaded here](https://console.apify.com/admin/users/wM4FzNZqJ8zKeBi7x/actors/6TUCqrM9GuOZsbYnp/source), for any queries, [join our discord community](https://apify.com/fis).
```

Comment:
Let's use GitHub to share the code instead of an Actor. The Actor's code won't be visible to people who aren't Apify admins 😅

Reply:
yeah yeah, posted it for you actually :D